### ISSUE TITLE
ANR on Support Lib 26

### ISSUE THREAD

**Issue by foodpoison** [googlesamples/androidtv-Leanback#98]
I discovered a ANR crash with Leanback that wasn't there in Lib 25, and can be easily reproduced in this demo app. Launch the app, select the "Demo Slam" row, and just held down "right" button, it'll spin into Activity Not Responding when it reaches the end. 
If you scroll slowly it's fine, it seems to happen on the end edges, and can happen on both left-most and right-most edges of a long enough row, most likely a recycling issue.

**Comment by benbaxter** [comment]
What device were you able to reproduce this on? I am attempting on a nvidia shield and not able to reproduce?

Can you provide more information or a stack trace?

**Comment by foodpoison** [comment]
The time that I reproduced this problem with this demo app I did it on a Xiaomi MiBox3. I couldn't reproduce this problem on the Nvidia Shield either with this demo app, but can with the app that I'm building which have a fairly large catalogue. My app breaks on the Nexus Player as well, though I don't currently have that device to test this demo app against, but at this point I'm fairly confident it's an issue with Support Library 26. 25 worked fine with all of these devices.

 I've made a low-res screen capture video of this happening on a low-end set top box (Amlogic TX3 Pro), and sharing it here on Dropbox: https://www.dropbox.com/s/wwxfzhu9c0qqegy/device-2017-10-05-143632.mp4?dl=0
(You'll see the screen freeze, then after a while the ANR dialog pops up. This video was not edited in anyway, just low-res).

In order to reproduce this issue on that box I needed to add a lot more content, basically in `MainFragment.java` L:266 I duplicated `mCategoryRowAdapter.add(row);` 4 times, so you'll see a lot more rows than was in the existing code, but contents/memory usage seems to be a factor.

And Logcat:
```
...
10-05 14:56:16.220 5761-5761/com.android.inputmethod.latin D/LatinIME: Trace_key, onKeyDown keyCode is:22
10-05 14:56:16.265 5761-5761/com.android.inputmethod.latin D/LatinIME: Trace_key, onKeyDown keyCode is:22
10-05 14:56:16.317 5761-5761/com.android.inputmethod.latin D/LatinIME: Trace_key, onKeyDown keyCode is:22
10-05 14:56:16.365 5761-5761/com.android.inputmethod.latin D/LatinIME: Trace_key, onKeyDown keyCode is:22
10-05 14:56:16.416 5761-5761/com.android.inputmethod.latin D/LatinIME: Trace_key, onKeyDown keyCode is:22
10-05 14:56:16.467 5761-5761/com.android.inputmethod.latin D/LatinIME: Trace_key, onKeyDown keyCode is:22
10-05 14:56:16.518 5761-5761/com.android.inputmethod.latin D/LatinIME: Trace_key, onKeyDown keyCode is:22
10-05 14:56:19.566 3638-4295/? D/audio_hw_primary: out_standby(0xf470f000)
10-05 14:56:19.566 3638-4295/? D/audio_hw_primary: do_output_standby(0xf470f000)
10-05 14:56:19.566 3638-4295/? I/audio_hw_primary: remove out (0xf470f000) from index 0
10-05 14:56:19.566 3638-4295/? I/audio_hw_primary: close pcm 0xf1fc2000
10-05 14:56:21.572 4350-5338/system_process I/InputDispatcher: Application is not responding: AppWindowToken{eee1417 token=Token{375b096 ActivityRecord{36603b1 u0 com.example.android.tvleanback/.ui.MainActivity t46}}} - Window{9e9027e u0 com.example.android.tvleanback/com.example.android.tvleanback.ui.MainActivity}.  It has been 5004.6ms since event, 5004.4ms since wait started.  Reason: Waiting to send key event because the focused window has not finished processing all of the input events that were previously delivered to it.  Outbound queue length: 0.  Wait queue length: 1.
10-05 14:56:21.574 4350-5338/system_process I/WindowManager: Input event dispatching timed out sending to com.example.android.tvleanback/com.example.android.tvleanback.ui.MainActivity.  Reason: Waiting to send key event because the focused window has not finished processing all of the input events that were previously delivered to it.  Outbound queue length: 0.  Wait queue length: 1.
10-05 14:56:21.634 4350-4397/system_process I/Process: Sending signal. PID: 7255 SIG: 3
10-05 14:56:21.635 7255-7260/com.example.android.tvleanback I/art: Thread[2,tid=7260,WaitingInMainSignalCatcherLoop,Thread*=0xef34e000,peer=0x12c740a0,"Signal Catcher"]: reacting to signal 3
10-05 14:56:21.839 4350-4397/system_process I/Process: Sending signal. PID: 4350 SIG: 3
10-05 14:56:21.840 4350-4359/system_process I/art: Thread[2,tid=4359,WaitingInMainSignalCatcherLoop,Thread*=0xef34b000,peer=0x12c010a0,"Signal Catcher"]: reacting to signal 3
10-05 14:56:21.847 7255-7260/com.example.android.tvleanback I/art: Wrote stack traces to '/data/anr/traces.txt'
10-05 14:56:21.850 4350-4397/system_process I/Process: Sending signal. PID: 5804 SIG: 3
10-05 14:56:21.850 5804-5865/com.droidlogic.dig I/art: Thread[2,tid=5865,WaitingInMainSignalCatcherLoop,Thread*=0xef34b000,peer=0x12c490a0,"Signal Catcher"]: reacting to signal 3
10-05 14:56:22.088 5804-5865/com.droidlogic.dig I/art: Wrote stack traces to '/data/anr/traces.txt'
10-05 14:56:22.895 4350-4397/system_process I/Process: Sending signal. PID: 5793 SIG: 3
10-05 14:56:22.896 5793-5842/com.droidlogic.SubTitleService I/art: Thread[2,tid=5842,WaitingInMainSignalCatcherLoop,Thread*=0xef34b000,peer=0x12c480a0,"Signal Catcher"]: reacting to signal 3
10-05 14:56:22.897 4350-4397/system_process I/Process: Sending signal. PID: 5786 SIG: 3
10-05 14:56:22.898 5786-5857/com.droidlogic.service.remotecontrol I/art: Thread[2,tid=5857,WaitingInMainSignalCatcherLoop,Thread*=0xef34b000,peer=0x12c430a0,"Signal Catcher"]: reacting to signal 3
10-05 14:56:23.009 5786-5857/com.droidlogic.service.remotecontrol I/art: Wrote stack traces to '/data/anr/traces.txt'
10-05 14:56:23.009 4350-4397/system_process I/Process: Sending signal. PID: 5781 SIG: 3
10-05 14:56:23.009 5781-5825/com.droidlogic.PPPoE I/art: Thread[2,tid=5825,WaitingInMainSignalCatcherLoop,Thread*=0xef34b000,peer=0x12c3f0a0,"Signal Catcher"]: reacting to signal 3
10-05 14:56:23.039 5793-5842/com.droidlogic.SubTitleService I/art: Wrote stack traces to '/data/anr/traces.txt'
10-05 14:56:23.039 4350-4397/system_process I/Process: Sending signal. PID: 5776 SIG: 3
10-05 14:56:23.039 5776-5824/com.droidlogic.inputmethod.remote I/art: Thread[2,tid=5824,WaitingInMainSignalCatcherLoop,Thread*=0xef34b000,peer=0x12c3b0a0,"Signal Catcher"]: reacting to signal 3
10-05 14:56:23.158 5776-5824/com.droidlogic.inputmethod.remote I/art: Wrote stack traces to '/data/anr/traces.txt'
10-05 14:56:23.158 4350-4397/system_process I/Process: Sending signal. PID: 5507 SIG: 3
10-05 14:56:23.159 5507-5539/com.android.systemui I/art: Thread[2,tid=5539,WaitingInMainSignalCatcherLoop,Thread*=0xef34b000,peer=0x12c350a0,"Signal Catcher"]: reacting to signal 3
10-05 14:56:23.217 5781-5825/com.droidlogic.PPPoE I/art: Wrote stack traces to '/data/anr/traces.txt'
10-05 14:56:23.270 4350-5454/system_process D/NetUtils: dhcp_start failed : eth0
10-05 14:56:23.388 5507-5539/com.android.systemui I/art: Wrote stack traces to '/data/anr/traces.txt'
10-05 14:56:24.378 4350-5454/system_process E/EthernetNetworkFactory: DHCP request error:Timed out waiting for DHCP to finish
10-05 14:56:24.399 4350-4359/system_process I/art: Wrote stack traces to '/data/anr/traces.txt'
10-05 14:56:25.297 4350-4397/system_process E/ActivityManager: ANR in com.example.android.tvleanback (com.example.android.tvleanback/.ui.MainActivity)
                                                               PID: 7255
                                                               Reason: Input dispatching timed out (Waiting to send key event because the focused window has not finished processing all of the input events that were previously delivered to it.  Outbound queue length: 0.  Wait queue length: 1.)
                                                               Load: 3.48 / 2.4 / 1.0
                                                               CPU usage from 145861ms to 0ms ago:
                                                                 69% 6326/iflix.sit1.play: 59% user + 10% kernel / faults: 6857 minor 46 major
                                                                 39% 5488/android.process.media: 29% user + 9.8% kernel / faults: 51960 minor 25 major
                                                                 5.7% 5487/sdcard: 0.9% user + 4.8% kernel / faults: 430 minor
                                                                 5.3% 4350/system_server: 3.1% user + 2.1% kernel / faults: 32942 minor 206 major
                                                                 1.8% 6189/com.google.android.gms: 1.1% user + 0.6% kernel / faults: 39623 minor 348 major
                                                                 1.6% 6136/adbd: 0.1% user + 1.5% kernel / faults: 1103 minor
                                                                 1.6% 6076/com.google.android.gms.persistent: 1.1% user + 0.5% kernel / faults: 21283 minor 110 major
                                                                 1.6% 2581/mmcqd/0: 0% user + 1.6% kernel
                                                                 1.4% 6245/ksdioirqd/sdio: 0% user + 1.4% kernel
                                                                 1.2% 1640/kswapd0: 0% user + 1.2% kernel
                                                                 0.8% 2409/irq/248-sd_emmc: 0% user + 0.8% kernel
                                                                 0.6% 2406/irq/250-sd_emmc: 0% user + 0.6% kernel
                                                                 0.6% 3627/surfaceflinger: 0.3% user + 0.3% kernel / faults: 971 minor 5 major
                                                                 0.5% 6360/RTW_CMD_THREAD: 0% user + 0.5% kernel
                                                                 0.5% 16/ksoftirqd/2: 0% user + 0.5% kernel
                                                                 0.4% 6271/com.facebook.katana: 0.3% user + 0.1% kernel / faults: 11798 minor 79 major
                                                                 0.3% 3638/mediaserver: 0.1% user + 0.1% kernel / faults: 3008 minor 61 major
                                                                 0.3% 1507/kworker/0:1: 0% user + 0.3% kernel
                                                                 0.3% 5812/com.droidlogic.mboxlauncher: 0.1% user + 0.1% kernel / faults: 13343 minor 68 major
                                                                 0.2% 6020/com.google.process.gapps: 0.1% user + 0.1% kernel / faults: 5557 minor 5 major
                                                                 0.2% 6359/RTW_XMIT_THREAD: 0% user + 0.2% kernel
                                                                 0.2% 2344/kworker/1:1: 0% user + 0.2% kernel
                                                                 0.2% 6361/RTWHALXT: 0% user + 0.2% kernel
                                                                 0.2% 20/ksoftirqd/3: 0% user + 0.2% kernel
                                                                 0.1% 7/rcu_preempt: 0% user + 0.1% kernel
                                                                 0.1% 3569/logd: 0% user + 0.1% kernel / faults: 255 minor
                                                                 0.1% 2482/kthread_di: 0% user + 0.1% kernel
                                                                 0% 5769/com.android.vending: 0% user + 0% kernel / faults: 4009 minor 52 major
                                                                 0% 3625/lmkd: 0% user + 0% kernel / faults: 36 minor
                                                                 0% 3/ksoftirqd/0: 0% user + 0% kernel
                                                                 0% 6/kworker/u8:0: 0% user + 0% kernel
                                                                 0% 3645/zygote: 0% user + 0% kernel / faults: 12378 minor
                                                                 0% 6247/wpa_supplicant: 0% user + 0% kernel / faults: 222 minor
                                                                 0% 12/ksoftirqd/1: 0% user + 0% kernel
                                                                 0% 24/kworker/u8:1: 0% user + 0% kernel
                                                                 0% 3557/jbd2/mmcblk0p14: 0% user + 0% kernel
                                                                 0% 5507/com.android.systemui: 0% user + 0% kernel / faults: 4216 minor 17 major
                                                                 0% 6331/com.google.android.youtube: 0% user + 0% kernel / faults: 2462 minor 21 major
                                                                 0% 3576/kworker/1:1H: 0% user + 0% kernel
                                                                 0% 1//init: 0% user + 0% kernel / faults: 377 minor
                                                                 0% 5761/com.android.inputmethod.latin: 0% user + 0% kernel / faults: 2208 minor 8 major
                                                                 0% 6921/kworker/2:2: 0% user + 0% kernel
                                                                 0% 2360/cpu_hotplug_gdb: 0% user + 0% kernel
                                                                 0% 3626/servicemanager: 0% user + 0% kernel / faults: 82 minor
                                                                 0% 6040/android.process.acore: 0% user + 0% kernel / faults: 1529 minor 3 major
                                                                 0% 2392/kthread_hdcp: 0% user + 0% kernel
                                                                 0% 3787/aml_buf_toggle: 0% user + 0% kernel
                                                                 0% 1634/kworker/0:1H: 0% user + 0% kernel
                                                                 0% 3635/netd: 0% user + 0% kernel / faults: 249 minor
                                                                 0% 3639/installd: 0% user + 0% kernel / faults: 211 minor 1 major
                                                                 0% 6531/com.android.deskclock: 0% user + 0% kernel / faults: 1412 minor 2 major
                                                                 0% 10/migration/0: 0% user + 0% kernel
                                                                 0% 15/migration/2: 0% user + 0% kernel
                                                                 0% 19/migration/3: 0% user + 0% kernel
                                                                 0% 2414/vmalloc_ion: 0% user + 0% kernel
                                                                 0% 2702/ueventd: 0% user + 0% kernel / faults: 25 minor
                                                                 0% 5781/com.droidlogic.PPPoE: 0% user + 0% kernel / faults: 294 minor 2 major
                                                                +0% 7019/kworker/3:2: 0% user + 0% kernel
                                                                +0% 7026/com.android.providers.calendar: 0% user + 0% kernel
                                                                +0% 7068/com.android.gallery3d: 0% user + 0% kernel
                                                                +0% 7084/com.android.launcher: 0% user + 0% kernel
                                                                +0% 7144/com.android.de
10-05 14:56:25.573 4350-4364/system_process I/art: Background sticky concurrent mark sweep GC freed 57519(3MB) AllocSpace objects, 17(1484KB) LOS objects, 23% free, 13MB/17MB, paused 41.970ms total 264.255ms
10-05 14:56:25.662 4350-7451/system_process D/OpenGLRenderer: Use EGL_SWAP_BEHAVIOR_PRESERVED: false
10-05 14:56:25.666 4350-4398/system_process W/ViewRootImpl: Cancelling event due to no window focus: KeyEvent { action=ACTION_UP, keyCode=KEYCODE_DPAD_RIGHT, scanCode=106, metaState=0, flags=0x28, repeatCount=0, eventTime=204542, downTime=201547, deviceId=4, source=0x301 }
10-05 14:56:25.666 4350-4398/system_process W/ViewRootImpl: Cancelling event due to no window focus: KeyEvent { action=ACTION_UP, keyCode=KEYCODE_DPAD_RIGHT, scanCode=106, metaState=0, flags=0x28, repeatCount=0, eventTime=204542, downTime=201547, deviceId=4, source=0x301 }
...
```

**Comment by foodpoison** [comment]
Hi @benbaxter I've got a fork which seems we can more reliably trigger this issue: https://github.com/foodpoison/androidtv-Leanback

Tested and can trigger the crash on Nexus Player running 8.0.0. Also tried on Nvidia Shield but harder to trigger it, seems like being on a higher-end device makes a difference, crashes consistently on lower-end devices, seems like a race-condition when the recyclerview reaches the end. Problem goes away if we move support library version back to 25.x.

**Comment by dakegu2** [comment]
please paste stacktrace of the app in /data/anr/traces.txt

**Comment by benbaxter** [comment]
@foodpoison This was indeed an issue with GridLayoutManager in leanback. We have a fix for it going out with the 27.0.0 support library.

Thank you for the detailed investigation!

I am going to close this issue and if it still persists after upgrading to support lib v27, then we can re-open it.

**Comment by foodpoison** [comment]
@dakegu2 ANR stacktrace linked [HERE](https://pastebin.com/raw/C5t3Ky4P)

**Comment by martinbonnin** [comment]
Same problem here. This seems to be an endless loop in `GridLayoutManager.layoutChildren`. 

I would guess somewhere around here: https://android.googlesource.com/platform/frameworks/support/+/9106804a84e5e8733e0b9313f749fa1f726e5d11/v17/leanback/java/src/android/support/v17/leanback/widget/GridLayoutManager.java#1014

**Comment by dakegu2** [comment]
Yes, the anr file looks like it's the bug we fixed internally, will be included in upcoming 27.0.0 release.

**Comment by martinbonnin** [comment]
27.0.0 works fine for us, Thanks :+1:  !!
